piston/utilities/format_output.py

```
@@ -0,0 +1,74 @@
+import json
+from typing import Optional
+
+import requests
+from piston.utilities.constants import FormatOutput
+
+
+def format_output(output: str) -> str:
+    """
+    Format the output and return a tuple of the formatted output and a URL to the full output.
+
+    Prepend each line with a line number. Truncate if there are over MAX_LINES lines or
+    MAX_CHARACTERS characters and upload the full output to a paste service.
+    """
+    output = output.strip()
+    original_output = output  # To be uploaded to a pasting service if needed
+    paste_link = None
+    truncated = False
+    lines = output.count("\n")
+
+    if lines > 0:
+        output = [f"{i:02d} | {line}" for i, line in enumerate(output.split("\n"), 1)]
+        output = output[: FormatOutput.MAX_LINES]
+        output = "\n".join(output)
+
+    if lines > FormatOutput.MAX_LINES:  # Limiting to only 20 lines
+        truncated = True
+        if len(output) >= FormatOutput.MAX_CHARACTERS:
+            output = f"{output[:FormatOutput.MAX_CHARACTERS]}\n... (truncated - too long, too many lines)"
+        else:
+            output = f"{output}\n... (truncated - too many lines)"
+    elif len(output) >= FormatOutput.MAX_CHARACTERS:
+        truncated = True
+        output = f"{output[:FormatOutput.MAX_CHARACTERS]}\n... (truncated - too long)"
+
+    if truncated:
+        paste_link = upload_output(original_output)
+
+    output = output or "[No output]"
+
+    msg = f"{output}\n"
+    if paste_link:
+        msg = f"{msg}\nFull output: {paste_link}"
```

We should add an `else` statement to indicate somewhere that pasting failed rather than skipping it entirely.